Remove unnecessary join calls

treeowl opened the pull request:
A couple spots used `Control.Monad.join` in ways that were
more confusing than useful. In particular, they used STM
to calculate an IO action, when in fact they simply needed
to run an STM action and then use the result more directly.
Not a big deal, but this way seems clearer.

treeowl committed:
Remove unnecessary join calls

A couple spots used `Control.Monad.join` in ways that were
more confusing than useful. In particular, they used STM
to calculate an IO action, when in fact they simply needed
to run an STM action and then use the result more directly.
Not a big deal, but this way seems clearer.

treeowl committed:
Be more explicit about bootstrap state changes

treeowl committed:
Rework bootstrapping

* Remove a gap in the exception handling.

* Use an asynchronous request for the initial ping.

* Set `BootstrapDone` on success.

* If bootstrapping is in progress when `bootstrap` is called,
  wait for it to complete and don't re-bootstrap if successful.

treeowl commented on src/Network/PeerDiscovery/Operations.hs:
```diff
@@ -68,9 +105,33 @@ bootstrap pd node = do
+    -- the ability to cancel requests, and I don't know if it's worth the
+    -- trouble.
+    initial_ping_succeeded <- readMVar initialPingMV
+    if not initial_ping_succeeded
```
Swap this test. Originally the `else` branch was much longer.

treeowl committed:
Reset public port on bootstrap failure

* Reset the public port if bootstrapping fails altogether.

* Implement the apparent intention to return `False` when an
  exception interrupts bootstrapping, instead of propagating
  the exception. Is this the right move? Not sure.

treeowl commented:
For now, this implements the (apparent) intention of `bootstrap` returning `False` if an exception is received. But this makes me quite nervous. Normally, one *shouldn't* just catch all exceptions. @arybczak, in what situation would `bootstrap` receive an exception that *shouldn't* propagate up to its caller?

arybczak commented:
> Normally, one shouldn't just catch all exceptions. @arybczak, in what situation would bootstrap receive an exception that shouldn't propagate up to its caller?

In none. The intention was to just reset the bootstrap state TVar when exception is received, so usage of `onException` was/is fine.

I skimmed over the code, it seems to be fine. I'll give a closer look shortly.

avieth commented on src/Network/PeerDiscovery/Operations.hs:
```diff
@@ -26,29 +26,81 @@ import Network.PeerDiscovery.Routing
+    resetBootstrap = do
+      modifyMVarP_ (pdPublicPort pd) (const requestedPublicPort)
+      atomically $ writeTVar (pdBootstrapState pd) BootstrapNeeded
+  (\(_ :: SomeException) -> False <$ resetBootstrap) `handle` do
```
Should re-throw this, no?

avieth reviewed (commented):
General question about bootstrapping: do we even need the STM concurrency control? Would it be feasible to not leave it up to the user to call bootstrap, and ensure that we only call it once, never concurrently?

treeowl committed:
Go back to propagating exceptions

I misunderstood the intent (thankfully). Go back to propagating
exceptions as usual.

treeowl commented:
@avieth, good question. The intention (based on `peer-discovery.md`) seems to be to potentially re-bootstrap if we detect that the network is tainted. But it may very well make more sense to shut down the `PeerDiscovery` instance in that case and leave it to the caller to start over from scratch. @arybczak, what do you think?

treeowl commented:
@avieth, actually.... I'm not sure that's the way to go. We don't usually want the caller to have to worry too much about what's going on under the hood. @arybczak, in the case of re-bootstrapping, do we want to clear out the routing table?

arybczak commented:
> @arybczak, in the case of re-bootstrapping, do we want to clear out the routing table?

Generally we would like to re-bootstrap if `findPeers` returns nothing, but the thing is that it may do that either when network was tainted or when internet connection is down. Ideally we'd want to clear the routing table in the former case, but not the latter.

avieth commented:
> it may very well make more sense to shut down the PeerDiscovery instance in that case and leave it to the caller to start over from scratch.
> ...
> actually.... I'm not sure that's the way to go. We don't usually want the caller to have to worry too much about what's going on under the hood. 

I think that, at least for cardano-sl, this _is_ a good way to go.

dcoutts commented on src/Network/PeerDiscovery/Operations.hs:
```diff
@@ -26,36 +26,73 @@ import Network.PeerDiscovery.Routing
+    BootstrapNeeded -> True <$ writeTVar (pdBootstrapState pd) BootstrapInProgress
+  if not bootstrap_still_needed
+  then return True
+  else do
```
Perhaps this is a good place to split it again, with the rest in a `bootstrapUnconditional` helper (or think of a better name).

dcoutts commented on src/Network/PeerDiscovery/Operations.hs:
```diff
@@ -68,9 +105,33 @@ bootstrap pd node = do
+          -- If we're not, erase the port so we don't pretend in future
+          -- communication that we are.
+          when (not reachable) $ do
+            putStrLn $ "bootstrap: we are not globally reachable on " ++ show port
```
We'll eventually have to tidy up logging like this to use a proper logger interface.

dcoutts commented on src/Network/PeerDiscovery/Operations.hs:
```diff
@@ -65,76 +65,81 @@ bootstrapUnlessDone
+  -- Save the requested public port. If something goes wrong
+  -- (we can't ping the initial peer at all or we receive an exception)
+  -- we restore this.
+  requestedPublicPort <- readMVar (pdPublicPort pd)
```
Why do we need to do this? The `pdPublicPort` can start off as `Just port` and only gets reset to `Nothing` if we fail to get a reply after a timeout (or get an invalid reply). Why do we need this extra save restore?

dcoutts commented on src/Network/PeerDiscovery/Operations.hs:
```diff
@@ -26,29 +26,81 @@ import Network.PeerDiscovery.Routing
+  -- 2. However, if bootstrapping is currently *in progress*, we just want
+  -- to wait for it to succeed.
+  --
+  -- We originally used just one transaction:
```
This is helpful in a commit message to explain the change, but the comments we want to leave in for our future selves and other readers is the explanation of what we do now and why. Typically the history isn't that relevant (including here I think), though explaining why we don't do it some alternative way can be useful in explaining why we do something the way we do it.

So I'd adjust this comment to just talk about what we do now and why.

dcoutts commented on src/Network/PeerDiscovery/Operations.hs:
```diff
@@ -26,29 +26,81 @@ import Network.PeerDiscovery.Routing
-              putStrLn $ "bootstrap: we are not globally reachable on " ++ show port
-              modifyMVarP_ (pdPublicPort pd) (const Nothing)
+bootstrap pd node = do
+  -- Why do we have this transaction and then one in bootstrapUnlessDone?
```
I'd clarify this comment: this *STM* transaction.

dcoutts commented on src/Network/PeerDiscovery/Operations.hs:
```diff
@@ -26,29 +26,81 @@ import Network.PeerDiscovery.Routing
+  
+      -- We do this asynchronously: there's no need to wait for a response
+      -- to the initial ping before sending the one that checks for global
+      -- reachability.
```
I think we could better explain the overall structure here in the comments. We're doing two pings and waiting for both to complete. We do that asymmetrically with an async request for the first and a sync one for the second. One checks the peer exists / is online while the other checks if we're reachable from the outside world (not firewalled).

dcoutts commented on src/Network/PeerDiscovery/Operations.hs:
```diff
@@ -59,12 +111,35 @@ bootstrap pd node = join . atomically $ readTVar (pdBootstrapState pd) >>= \case
-    result st = do
-      atomically $ writeTVar (pdBootstrapState pd) BootstrapNeeded
-      return $ st == BootstrapDone
+        putMVar initialPingMV True
```
Check the exception handling here. If we get an exception in the response handler I presume the other handler for `sendRequest` does not get run. In which case here the `putMVar initialPingMV` would never happen and the main thread would hang in reading that MVar (or if it's lucky get a deadlocked on MVar exception).

We might want to add another util like the existing `sendRequestSync`, to use more symmetrically (and avoiding exception issues) something like:

```
newtype Async a = Async (MVar a)

sendRequestAsync
  :: (IsRequest req, Show req)
  => PeerDiscovery cm
  -> req                        -- ^ Request to be sent
  -> Node                       -- ^ Recipient
  -> IO (Async (ResponseType req))
sendRequestSync pd reqType peer = do
  v <- newEmptyMVar
  sendRequest pd reqType peer
    (putMVar v Nothing)
    (putMVar v . Just)
  return (Async v)

awaitAsync (Async v) = readMVar v
```
And then used like
```
  -- Send both ping requests. Though actually sending the second is conditional.
  pingAsync1 <- sendRequestAsync pd (Ping Nothing)
  pingAsync2 <- maybe (return Nothing)
                      (\port -> Just <$> sendRequestAsync pd (Ping (Just port)))
                      requestedPublicPort
  -- Wait for both responses (or failure). Again the second is conditional.
  ping1Res <- awaitAsync pingAsync1
  ping2Res <- maybe (return Nothing) (fmap Just . awaitAsync) pingAsync2

  case ping1Res of
    Nothing -> ...
    Just Pong -> ...

  case ping2Res of
    Nothing -> ... -- was not needed
    Just Nothing -> ... -- failure
    Just (Just Pong) -> ...
```

dcoutts reviewed (changes requested):
Looks good. A few suggestions to tidy things up and explain more.

When we're done we should either squash this into one commit, or squash our fixups. I don't mind which.
